### ISSUE TITLE
[Bug] Implementation of Eisner's algorithm does not restrict the root number to 1 

### ISSUE THREAD

**Issue by sustcsonglin** [harvardnlp/pytorch-struct#74]
Hey,  I found that your implementation of Eisner's algorithm admits arbitrary root number, which is a very severe bug since dependency parsing usually has only one root token.  

In your  DepTree._dp() method, you make a conversion to let the root token as the first token in the sentence.  Imagine that  the root x_{0} attacks word x_{i},   I_{0,0} + C_{1, i} = I_{0, i}  and  I_{0, i} + C_{i,j} = C_{0, j}  for some j < L where L is the length of sentence.  Now  complete span C_{0, j} still  have opportunity to attach a new word x_{k} for  j< k<=L, making multiple root attachment possible.

Fortunately, I made some changes to your codes to restrict the root number to 1.

`    def _dp(self, arc_scores_in, lengths=None, force_grad=False, cache=True):
        if arc_scores_in.dim() not in (3, 4):
            raise ValueError("potentials must have dim of 3 (unlabeled) or 4 (labeled)")

        labeled = arc_scores_in.dim() == 4
        semiring = self.semiring
        # arc_scores_in = _convert(arc_scores_in)
        arc_scores_in, batch, N, lengths = self._check_potentials(
            arc_scores_in, lengths
        )
        arc_scores_in.requires_grad_(True)
        arc_scores = semiring.sum(arc_scores_in) if labeled else arc_scores_in
        alpha = [
            [
                [
                    Chart((batch, N, N), arc_scores, semiring, cache=cache)
                    for _ in range(2)
                ]
                for _ in range(2)
            ]
            for _ in range(2)
        ]

        semiring.one_(alpha[A][C][L].data[:, :, :, 0].data)
        semiring.one_(alpha[A][C][R].data[:, :, :, 0].data)
        semiring.one_(alpha[B][C][L].data[:, :, :, -1].data)
        semiring.one_(alpha[B][C][R].data[:, :, :, -1].data)

        for k in range(1, N):
            f = torch.arange(N - k), torch.arange(k, N)
            ACL = alpha[A][C][L][: N - k, :k]
            ACR = alpha[A][C][R][: N - k, :k]
            BCL = alpha[B][C][L][k:, N - k :]
            BCR = alpha[B][C][R][k:, N - k :]
            x = semiring.dot(ACR, BCL)
            arcs_l = semiring.times(x, arc_scores[:, :, f[1], f[0]])
            alpha[A][I][L][: N - k, k] = arcs_l
            alpha[B][I][L][k:N, N - k - 1] = arcs_l
            arcs_r = semiring.times(x, arc_scores[:, :, f[0], f[1]])
            alpha[A][I][R][:N - k, k] = arcs_r
            alpha[B][I][R][k:N, N - k - 1] = arcs_r
            AIR = alpha[A][I][R][: N - k, 1 : k + 1]
            BIL = alpha[B][I][L][k:, N - k - 1 : N - 1]
            new = semiring.dot(ACL, BIL)
            alpha[A][C][L][: N - k, k] = new
            alpha[B][C][L][k:N, N - k - 1] = new
            new = semiring.dot(AIR, BCR)
            alpha[A][C][R][: N - k, k] = new
            alpha[B][C][R][k:N, N - k - 1] = new

        root_incomplete_span = semiring.times(alpha[A][C][L][0, :], arc_scores[:, :, torch.arange(N), torch.arange(N)])
        root =  [ Chart((batch,), arc_scores, semiring, cache=cache) for _ in range(N)]
        for k in range(N):
            AIR = root_incomplete_span[:, :, :k+1]
            BCR = alpha[B][C][R][k, N - (k+1):]
            root[k] = semiring.dot(AIR, BCR)
        v = torch.stack([root[l-1][:,i] for i, l in enumerate(lengths)], dim=1)
        return v, [arc_scores_in], alpha
`

Basically,  I don't treat the first token as root anymore.  I handle the root token just after the for-loop, so you may need handle the length variable. (length = length-1, root no longer be treated as part of sentence) .  I tested the modified code and found it bug-free

**Comment by sustcsonglin** [comment]
`       

        root_incomplete_span = semiring.times(alpha[A][C][L][0, :], arc_scores[:, :, torch.arange(N), torch.arange(N)])
        root =  [ Chart((batch,), arc_scores, semiring, cache=cache) for _ in range(N)]
        for k in range(N):
            AIR = root_incomplete_span[:, :, :k+1]
            BCR = alpha[B][C][R][k, N - (k+1):]
            root[k] = semiring.dot(AIR, BCR)


`

The key idea is to obtain the incomplete span I_{0,i} first, then make sure that I_{0,i} combines with C_{i, L} only, where L is the length of sentence.

**Comment by srush** [comment]
Thanks, can you send a pull request? 

This was done on purpose, as I needed the multi-root case. But I think having both options would be preferable.

**Comment by sustcsonglin** [comment]
I implement the single-root case, but I found a weird issue. The gradient is incorrect when I set the argument "cache=True" for Chart class, while when I set the argument to "cache=False", the gradient is correct.

**Comment by srush** [comment]
Hmm, let's turn off caching by default then. It gives a small speed up, but you need to be careful to ensure you only use each cell in the table once. 

Btw, if you are interested it would be really easy to add a test. This code can enumerate all single-root trees:

https://github.com/harvardnlp/pytorch-struct/blob/master/torch_struct/deptree.py#L178

You could add another test case here: 

https://github.com/harvardnlp/pytorch-struct/blob/master/torch_struct/test_algorithms.py#L170
